cockatrice/src/dlg_connect.cpp

```diff
@@ -156,26 +167,104 @@ void DlgConnect::actSaveConfig()
+        QStringList publicServersAvailable;
+        QStringList publicServersToRemove;
+
+        // Downloaded data from GitHub
+        QString jsonData = QString(reply->readAll());
+        auto jsonMap = QtJson::Json::parse(jsonData).toMap();
+
+        // Servers available
+        auto publicServersJSONList = jsonMap["servers"].toList();
+        for (const auto &server : publicServersJSONList) {
+            // Data inside one server at a time
+            // server: [{ ... }, ..., { ... }]
+            const auto serverMap = server.toMap();
+
+            QString serverName = serverMap["name"].toString();
+
+            if (serverMap["isInactive"].toBool()) {
+                publicServersToRemove.append(serverName);
+                continue;
+            }
 
+            QString serverAddress = serverMap["host"].toString();
+            QString serverPort = serverMap["port"].toString();
+
+            publicServersAvailable.append(serverName);
+
+            if (!savedHostList.contains(serverName)) {
+                settingsCache->servers().addNewServer(serverName, serverAddress, serverPort, "", "", false);
+            } else {
+                settingsCache->servers().updateExistingServerWithoutLoss(serverName, serverAddress, serverPort);
+            }
+        }
+
+        // If a server was removed from the public list,
+        // we will delete it from the local system.
+        // Will not delete "unofficial" servers
+        for (auto server : savedHostList) {
+            QString serverName = server.getSaveName();
+            bool isCustom = server.isCustomServer();
+
+            if (!isCustom && publicServersToRemove.indexOf(serverName) != -1) {
+                settingsCache->servers().removeServer(serverName);
+            }
+        }
+
+        reply->deleteLater();
+    } else {
+        qDebug() << "[CONNECTION DIALOG]"
+                 << "Error Downloading Public Servers" << errorCode;
+    }
+
+    // Signal that we're done with updating the combo box
+    emit sigPublicServersDownloaded();
+}
+
+void DlgConnect::preRebuildComboBoxList()
+{
     UserConnection_Information uci;
     savedHostList = uci.getServerInfo();
 
     if (savedHostList.size() == 1) {
-        settingsCache->servers().addNewServer("Rooster Ranges", "server.cockatrice.us", "4747", "", "", false);
-        settingsCache->servers().addNewServer("Chickatrice", "chickatrice.net", "4747", "", "", false);
-        settingsCache->servers().addNewServer("dr4ft", "cockatrice.dr4ft.com", "4747", "", "", false);
-        settingsCache->servers().addNewServer("Tetrarch", "mtg.tetrarch.co", "4747", "", "", false);
-    } else if (!savedHostList.contains("Rooster Ranges")) {
-        settingsCache->servers().addNewServer("Rooster Ranges", "server.cockatrice.us", "4747", "", "", false);
-        settingsCache->servers().removeServer("Woogerworks");
+        downloadPublicServers();
+    } else {
+        emit sigPublicServersDownloaded();
```

Comment:
If the saved host list != size 1, emit a signal saying we downloaded a server list? The logic doesn't make sense to me at all. What is this for?

Reply:
This signal just means continue the rebuilding process. If the size != 1 then we have info already stored locally and do not need to do a first time run install